drivers/dma/dma_stm32.c

```diff
@@ -0,0 +1,616 @@
+	}
+
+	ret = dma_stm32_get_memory_increment(memory_addr_adj,
+					&DMA_InitStruct.MemoryOrM2MDstIncMode);
+	if (ret < 0) {
+		return ret;
+	}
+	ret = dma_stm32_get_periph_increment(periph_addr_adj,
+					&DMA_InitStruct.PeriphOrM2MSrcIncMode);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (config->head_block->source_reload_en) {
+		DMA_InitStruct.Mode = LL_DMA_MODE_CIRCULAR;
+	} else {
+		DMA_InitStruct.Mode = LL_DMA_MODE_NORMAL;
+	}
+
+	source_periph = stream->direction == MEMORY_TO_PERIPHERAL;
+	ret = dma_stm32_width_config(config, source_periph, dma,
+				     &DMA_InitStruct, id);
+	if (ret < 0) {
+		return ret;
+	}
+	msize = DMA_InitStruct.MemoryOrM2MDstDataSize;
+
+#if defined(CONFIG_DMA_STM32_V1)
+	DMA_InitStruct.MemBurst = stm32_dma_get_mburst(config, source_periph);
+	DMA_InitStruct.PeriphBurst = stm32_dma_get_pburst(config,
+							source_periph);
+
+	if (config->channel_direction != MEMORY_TO_MEMORY) {
+		if (config->dma_slot >= 8) {
+			LOG_ERR("dma slot error.");
+			return -EINVAL;
+		}
+	} else {
+		if (config->dma_slot >= 8) {
+			LOG_ERR("dma slot is too big, using 0 as default.");
+			config->dma_slot = 0;
+		}
+	}
+	DMA_InitStruct.Channel = config->dma_slot;
+
+	stm32_dma_get_fifo_threshold(config->head_block->fifo_mode_control);
+
+	if (stm32_dma_check_fifo_mburst(&DMA_InitStruct)) {
+		DMA_InitStruct.FIFOMode = LL_DMA_FIFOMODE_ENABLE;
+	} else {
+		DMA_InitStruct.FIFOMode = LL_DMA_FIFOMODE_DISABLE;
+	}
+#endif
+	DMA_InitStruct.NbData = config->head_block->block_size;
+
+	LL_DMA_Init(dma, table_ll_stream[id], &DMA_InitStruct);
+
+	LL_DMA_EnableIT_TC(dma, table_ll_stream[id]);
+
+#if defined(CONFIG_DMA_STM32_V1)
+	if (DMA_InitStruct.FIFOMode == LL_DMA_FIFOMODE_ENABLE) {
+		LL_DMA_EnableIT_FE(dma, table_ll_stream[id]);
+	} else {
+		LL_DMA_DisableIT_FE(dma, table_ll_stream[id]);
+	}
+#endif
+	return ret;
+}
+
+int dma_stm32_disable_stream(DMA_TypeDef *dma, u32_t id)
+{
+	int count = 0;
+
+	for (;;) {
+		if (!stm32_dma_disable_stream(dma, id)) {
+		}
+		/* After trying for 5 seconds, give up */
+		if (count++ > (5 * 1000)) {
+			return -EBUSY;
+		}
+		k_sleep(K_MSEC(1));
+	}
+
+	return 0;
+}
+
+static int dma_stm32_reload(struct device *dev, u32_t id,
+			    u32_t src, u32_t dst, size_t size)
+{
+	const struct dma_stm32_config *config = dev->config->config_info;
+	DMA_TypeDef *dma = (DMA_TypeDef *)(config->base);
+	struct dma_stm32_data *data = dev->driver_data;
+	struct dma_stm32_stream *stream = &data->streams[id];
+
+	if (id >= data->max_streams) {
+		return -EINVAL;
+	}
+
+	switch (stream->direction) {
+	case MEMORY_TO_PERIPHERAL:
+		LL_DMA_SetMemoryAddress(dma, table_ll_stream[id], src);
+		LL_DMA_SetPeriphAddress(dma, table_ll_stream[id], dst);
+		break;
+	case MEMORY_TO_MEMORY:
+	case PERIPHERAL_TO_MEMORY:
+		LL_DMA_SetPeriphAddress(dma, table_ll_stream[id], src);
+		LL_DMA_SetMemoryAddress(dma, table_ll_stream[id], dst);
+		break;
+	default:
+		return -EINVAL;
+	}
+
```

Comment:
I think that here you need to set the size, isn't it?
Like:

LL_DMA_SetDataLength(dma, table_ll_stream[id], size);

Reply:
@cybertale 
I am now able to compile correctly, and I test it using I2S audio with microphone board.
But it does not work.
I tried to fix the dma_reload() function (see my latest review), but still does not work.
It seems that it stops receving data. Tomorrow I'll debug more.